File: project/src/Dungeon/Scripts/Dungeon.ts
import { Level } from "../../LevelCreator/Level";
import { DungeonNode, DNodes, DVoid, DTunnel, DEnemy, DExit, DGold } from "./DungeonNode";
import { Player } from "../../Object/Player";
import { QuestType } from "../../Object/Quest";

//PlaceHolder level
const LEVELNAME = ""
const EMPTY_LEVEL: Level = {
    name: "empty",
    difficulty: 0,
    map: [],
    start: [0, 0],
    exit: [4, 4],
    reward: {
        gold: 0
    },
    rows: 5,
    cols: 5
}
/**
 * Controls the Dungeon game
 */
export class Dungeon {
    private _player: [number, number]
    private _map: DungeonNode[][]
    readonly level: Level
    readonly exit: [number, number]
    readonly rows: number
    readonly cols: number
    /**
     * 
     * @param level a parseable array of numbers representing a level
     */
    constructor(levelname = LEVELNAME) {
        this.level = this.loadLevel(levelname)
        this._player = this.level.start
        this.exit = this.level.exit
        this.rows = this.level.rows
        this.cols = this.level.cols
        this._map = this.loadMap()
    }
    /**Current map of Dungeon Nodes */
    public get map() {
        return this._map
    }
    /**Current Player Coordinates*/
    public get player() {
        return this._player
    }
    /**
     * Moves the player to given coordinates if legal
     * @param row 
     * @param col 
     * @returns updated player coordinates if legal, otherwise original coordinates
     */
    public move(row: number, col: number) {
        if (this.canMove(row, col))
            this._player = [row, col]
        return this._player
    }

    public setMap(map: DungeonNode[][]) {
        if (this.rows === map.length && this.cols === map[0].length)
            this._map = map

    }
    /**
     * Instantiate DungeonNodes from level array
     */
    private loadMap(): DungeonNode[][] {
        if (!this.level) return []
        return this.level.map
            //Map number[] to DNode[]
            .map(row =>
                //Map numbers to DNodes based on the DNodes enum
                row.map(val => this.valToNode(val)))
    }

    private loadLevel(levelName: string) {
        let level = EMPTY_LEVEL
        try {
            level = window.api.readLevel(levelName)
        }
        catch (error) {
            console.log(error)
            console.log("Level is Empty")
        }
        console.log(level)
        return level
    }
    /**
     * Checks if movement to row, col is legal
     * @param row 
     * @param col 
     * @returns 
     */
    private canMove(row: number, col: number) {
        if (row < 0 || col < 0) return false
        if (row >= this.rows || col >= this.cols) return false
        return (!this._map[row][col].isVoid && !this.isComplete())
    }

    public isComplete() {
        return this._player[0] === this.exit[0] &&
            this._player[1] === this.exit[1]

    }

    private valToNode(val: number): DungeonNode {

        switch (val) {
            case DNodes.DVoid:
                return new DVoid();
            case DNodes.DTunnel:
                return new DTunnel();
            case DNodes.DExit:
                return new DExit();
            case DNodes.DGold:
                return new DGold()
            case DNodes.DEnemy:
                return new DEnemy();

            default: return new DVoid()
        }

    }

    public loseHealth(playerData: Player, setPlayerData: (player: Player) => void,
        acceptedQuests: QuestType[], processQuest: (quest: QuestType) => void) {

        let monsterHuntingQuest = acceptedQuests.find((quest) => quest.name === "MonsterHunting");
        let fingerSnappingQuest = acceptedQuests.find((quest) => quest.name === "FingerSnapping");

        let attack = 0;
        if (playerData.class === "Warrior") {
            attack = playerData.stats.Strength - 3 + playerData.stats.Vitality * 0.9;
        } else if (playerData.class === "Mage") {
            attack = playerData.stats.Intelligence + playerData.stats.Vitality * 0.2;
        } else if (playerData.class === "Ranger") {
            attack =
                (playerData.stats.Dexterity +
                    (playerData.stats.Luck * 0.4) +
                    (playerData.stats.Agility * 0.6)) / 2
                + playerData.stats.Vitality * 0.2;
        }

        let damage = Math.ceil(this.level.difficulty - attack / 3);
        console.log(damage);
        const newPlayerData = { ...playerData };
        const player_row = this._player[0];
        const player_col = this._player[1];
        if (this._map[player_row][player_col].color === "red") {

            if (newPlayerData.stats.HP >= damage) {
                newPlayerData.stats.HP -= damage;
            } else if (newPlayerData.stats.MaxHP > 0) {
                newPlayerData.stats.HP = 0;
            }
            console.log(newPlayerData.stats.HP);
            setPlayerData(newPlayerData);

            if (monsterHuntingQuest) {
                processQuest(monsterHuntingQuest);
            }
            if (fingerSnappingQuest) {
                processQuest(fingerSnappingQuest);
            }

        }
    }
}

File: project/public/python/Dungeon.py
import json
import random as rand
class Game:
    def __init__(self, board_file, stats):
        level = json.loads(board_file)
        playerstats = json.loads(stats)
        self.board = level["map"]
        self.rows = level["rows"]
        self.cols = level["cols"]
        self.difficulty = level["difficulty"]
        self.gold_reward = level["reward"]
        self.player_row = level["start"][0]
        self.player_col = level["start"][1]
        self.boardfile_name = level["name"]
        self.exit = level["exit"]
        self.win = False
        self.reward_count = 0
        self.player_hp = int(playerstats["HP"])
        self.player_mp = int(playerstats["MP"])
        self.player_max_hp = int(playerstats["MaxHP"])
        self.player_max_mp = int(playerstats["MaxMP"])
        self.player_level = int(playerstats["Level"])
        self.player_str = int(playerstats["Strength"])
        self.player_dex = int(playerstats["Dexterity"])
        self.player_luck = int(playerstats["Luck"])
        self.player_int = int(playerstats["Intellegence"])
        self.player_vit = int(playerstats["Vitality"])
        self.player_agil = int(playerstats["Agility"])
        
    def print_board(self):
        for row in self.board:
          for val in row:
            print(val, end=" ")
          print()
    def move(self, row, col):
        if (self.canMove(row, col)):
            self.player_row = row
            self.player_col = col
            if self.board[self.player_row][self.player_col] == 4 or self.board[self.player_row][self.player_col] == 6 or self.board[self.player_row][self.player_col] == 5:
              self.player_hp = self.player_hp - 1
              self.player_mp -= 1
        return (self.player_row, self.player_col)

    def canMove(self, row, col):
        if (row < 0 or col < 0):
          return False
        if (row >= self.rows or col >= self.cols): 
          return False
        return (self.board[row][col]!= 0 and not self.player_win())

    def get_player_pos(self):
        return (self.player_row, self.player_col)
  
    def player_win(self):
      if self.player_row == self.exit[0] and self.player_col == self.exit[1] and self.player_hp > 0:
        self.win = True
        self.reward_count = rand.randint(int(self.rows * self.cols),int(self.rows * self.cols * self.difficulty))
        self.player_max_hp += rand.randint(1, self.difficulty)
        self.player_max_mp += rand.randint(1, self.difficulty)
        self.player_level += rand.randint(1, self.difficulty)
        self.player_str += rand.randint(1, self.difficulty)
        self.player_dex += rand.randint(1, self.difficulty)
        self.player_luck += rand.randint(1, self.difficulty)
        self.player_int += rand.randint(1, self.difficulty)
        self.player_vit += rand.randint(1, self.difficulty)
        self.player_agil += rand.randint(1, self.difficulty)
        return True
      else:
        self.reward_count = -rand.randint(1,int(self.rows * self.cols * self.difficulty))
        return False

game = Game(f"{filename}", f"{newstats}")